File: backend/app/services/document_service.py
"""Document application logic over persisted prototype drafts."""

from fastapi import HTTPException, status
from sqlalchemy.exc import SQLAlchemyError
from sqlalchemy.orm import Session

from app.models.documents import Document
from app.repositories import document_repository
from app.schemas.documents import (
    DocumentDraftRequest,
    DocumentDraftResponse,
    DocumentTemplateResponse,
    DocumentType,
)

_TEMPLATE_TITLES = {
    DocumentType.RENTAL: "Rental Agreement",
    DocumentType.EMPLOYMENT: "Employment Agreement",
    DocumentType.NDA: "Non-Disclosure Agreement",
    DocumentType.WILL: "Will / Testament",
}


def _format_date(document: Document) -> str:
    if document.created_at is None:
        return "Just now"
    return document.created_at.strftime("%b %d, %Y")


def list_templates() -> list[DocumentTemplateResponse]:
    """Return the prototype document templates."""
    return [
        DocumentTemplateResponse(
            type=item["type"],
            title=item["title"],
            description=item["description"],
            category=item["category"],
        )
        for item in document_repository.list_templates()
    ]


def list_documents(db: Session, user_id: str) -> list[DocumentDraftResponse]:
    """Return persisted prototype drafts belonging to the authenticated user."""
    return [
        _to_response(document)
        for document in document_repository.list_documents(db, user_id)
    ]


def get_document(
    db: Session, document_id: str, user_id: str
) -> DocumentDraftResponse:
    """Return a single prototype draft owned by user or raise 404."""
    document = document_repository.get_document(db, document_id, user_id)
    if document is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Document not found.",
        )
    return _to_response(document)


def create_draft(
    db: Session, user_id: str, payload: DocumentDraftRequest
) -> DocumentDraftResponse:
    """Create a user-owned prototype draft."""
    title = f"{_TEMPLATE_TITLES[payload.type]} Draft"
    try:
        document = document_repository.create_document(
            db, user_id, payload.type.value, title, dict(payload.details)
        )
        db.commit()
        db.refresh(document)
    except SQLAlchemyError as exc:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Could not create the prototype draft.",
        ) from exc
    return _to_response(document)


def _to_response(document: Document) -> DocumentDraftResponse:
    return DocumentDraftResponse(
        document_id=document.id,
        type=DocumentType(document.type),
        title=document.title,
        status=document.status,
        created_date=_format_date(document),
        details={
            str(key): str(value)
            for key, value in dict(document.details or {}).items()
        },
        prototype=True,
    )

File: frontend/src/services/documentService.ts
/**
 * Document service communicating with the authenticated FastAPI backend.
 */

import { apiRequest } from './apiClient'
import {
  TEMPLATE_FORM_SCHEMAS,
  type DocumentDraft,
  type DocumentDraftRequest,
  type DocumentDraftResponse,
  type DocumentTemplate,
  type DocumentTemplateResponse,
  type DocumentType,
} from '../types/document'

function mapDocumentResponse(dto: DocumentDraftResponse): DocumentDraft {
  return {
    id: dto.document_id,
    type: dto.type,
    title: dto.title,
    createdDate: dto.created_date,
    status: dto.status,
    values: dto.details || {},
  }
}

const DEFAULT_TEMPLATES_FALLBACK: Record<
  DocumentType,
  { title: string; description: string; category: string }
> = {
  rental: {
    title: 'Rental Agreement',
    description:
      'Create a structured rental agreement draft from property and tenancy details.',
    category: 'Tenancy',
  },
  employment: {
    title: 'Employment Agreement',
    description:
      'Prepare a structured employment agreement draft from role and employment details.',
    category: 'Employment',
  },
  nda: {
    title: 'Non-Disclosure Agreement',
    description:
      'Create a structured NDA draft from the parties and confidentiality requirements.',
    category: 'Confidentiality',
  },
  will: {
    title: 'Will / Testament',
    description: 'Organize basic information for a will or testament draft.',
    category: 'Estate',
  },
}

export async function getDocumentTemplates(): Promise<DocumentTemplate[]> {
  try {
    const dtos = await apiRequest<DocumentTemplateResponse[]>(
      '/documents/templates',
    )
    return dtos.map((dto) => {
      const schema = TEMPLATE_FORM_SCHEMAS[dto.type] || {
        fields: [],
        sections: [],
      }
      return {
        type: dto.type,
        title: dto.title,
        description: dto.description,
        category: dto.category,
        fields: schema.fields,
        sections: schema.sections,
      }
    })
  } catch {
    return (['rental', 'employment', 'nda', 'will'] as DocumentType[]).map(
      (type) => getDocumentTemplate(type),
    )
  }
}

export function getDocumentTemplate(
  type: DocumentType,
  templates?: DocumentTemplate[],
): DocumentTemplate {
  if (templates && templates.length > 0) {
    const match = templates.find((item) => item.type === type)
    if (match) return match
  }
  const fallback = DEFAULT_TEMPLATES_FALLBACK[type]
  const schema = TEMPLATE_FORM_SCHEMAS[type] || { fields: [], sections: [] }
  return {
    type,
    title: fallback ? fallback.title : type,
    description: fallback ? fallback.description : '',
    category: fallback ? fallback.category : '',
    fields: schema.fields,
    sections: schema.sections,
  }
}

export async function getDocuments(): Promise<DocumentDraft[]> {
  const dtos = await apiRequest<DocumentDraftResponse[]>('/documents')
  return dtos.map(mapDocumentResponse)
}

export async function getDocument(id: string): Promise<DocumentDraft> {
  const dto = await apiRequest<DocumentDraftResponse>(`/documents/${id}`)
  return mapDocumentResponse(dto)
}

export async function createDraft(
  type: DocumentType,
  values: Record<string, string>,
): Promise<DocumentDraft> {
  const payload: DocumentDraftRequest = {
    type,
    details: values,
  }
  const dto = await apiRequest<DocumentDraftResponse>('/documents', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
  return mapDocumentResponse(dto)
}
